Add Arrow + GenKnowSub to LoRA

TheTahaaa opened the pull request:
Hi there! 👋

This PR adds support for **Arrow**, a modular routing mechanism for LoRA experts introduced [here](https://arxiv.org/abs/2405.11157), as well as our refinement method **GenKnowSub**, proposed in our ACL 2025 Main Conference [paper](https://arxiv.org/abs/2505.10939). GenKnowSub enhances Arrow by subtracting a general-domain LoRA from task-specific ones prior to routing, leading to improved generalisation and modularity.

The integration is implemented through a new ArrowLoraLinearLayer, registered as a LoraVariant (similar to DoRA).

### 🔧 Code Changes

Modified files under peft/tuners/lora/:

- config.py, layer.py, model.py, variants.py
- bnb.py (adds support for Arrow with bitsandbytes quantisation)

Added a new file:

- arrow.py – contains the core logic for Arrow and GenKnowSub

### ✅ Testing & Quality
Confirmed that tests/test_lora_variants.py passes.

All formatting and style checks pass (make quality successful).

### 📚 Optional Follow-up
I’ve also developed a separate GitHub repository demonstrating how to use Arrow and GenKnowSub on benchmark datasets (e.g., BoolQ, PIQA, ARC-Challenge). I’d be happy to contribute this as an example under peft/examples/ if it’s of interest.

Thank you in advance for reviewing this PR!

githubnemo commented:
Hey @TheTahaaa, thank you for the PR!

I'm unsure about the use of a LoRA variant here. The concept of Arrow is, as I understand it, more similar to [X-Lora](https://huggingface.co/docs/peft/main/en/package_reference/xlora) (i.e. a routing of multiple adapters) than a single adapter. Therefore I think it would make more sense to implement it as its own method (but still being able to ingest LoRA adapters, like X-LoRA does).

Secondly, I think that while a direct integration of `GenKnowSub` via a method can have certain utility, it seems more like a recipe or pre-processing step rather than something that should be available all the time. Maybe it would be more informative to have a full-fledged example how to create a general knowledge LoRA and how to subtract it?

Let me know what you think.

githubnemo commented:
> My understanding is that most of the work will involve refactoring existing code into a new class for the standalone Arrow method – is that correct?

Not quite what I meant. My suggestion is to keep the variant solution that you already have because it re-uses a lot of code from LoRA and making it its own method would duplicate a lot of code. To minimize the impact on future configs I suggested to use a breakout config (point (1) in my post). Since the variant implementation will not have a `PeftModel` class that checks user inputs related to Arrow/GenKnowSub I suggested (point (2) in my post) to implement a helper function that takes a model, adapters and whatever is needed to compile an arrow routing setup (possibly with GenKnowSub) and builds a model similar to what you provided in your example.

So instead of this:

```python

# Load task-specific LoRA adapters
model.load_adapter(..., adapter_name="cluster0")
...
model.load_adapter(..., adapter_name="cluster9")

# Load general-domain LoRA adapters (e.g., English, French, German)
model.load_adapter(..., adapter_name="le_en")
model.load_adapter(..., adapter_name="le_fr")
model.load_adapter(..., adapter_name="le_de")

# Now the model contains all the task-specific and general-domain LoRAs

# Define Arrow + GenKnowSub config as a dummy LoRA router
router_cfg = LoraConfig(
    r=2, # dummy rank since A and B won't be used!
    use_arrow=True, # This will turn this LoRA to the ArrowLoraVariant
    arrow_expert_num=10, # Number of task-specific modules in each LoRA layer
    arrow_top_k=3, # Number of selected task-specific LoRAs for each token in each layer 
    arrow_router_temperature=1.0,
    use_gks=True,  # ← enable GenKnowSub!
    le_names=..., # name of loaded general-domain LoRAs
    ts_names=..., # name of loaded task-specific LoRAs
    target_modules=["qkv_proj", "o_proj"]
)

# Add a dummy router adapter to trigger Arrow + GenKnowSub logic
model.add_adapter(adapter_name="router", peft_config=router_cfg)
model.set_adapter("router")

# Done! Just call `model.generate()` or `model(...)` as usual.
```

there will be something like

```python
from peft import create_arrow_model, ArrowConfig

peft_model = create_arrow_model(
    base_model=base_model,
    task_specific_adapters=[path0, path1, path2],
    general_knowledge_adapters=[path3, path4, path5],
    arrow_config=ArrowConfig(...)
)
```

This function would live in the `lora/variants.py` file and could check if the adapters are compatible (e.g., equally ranked).

TheTahaaa commented:
That would be much more straightforward to implement, working on it! @githubnemo

TheTahaaa commented:
✅ Implementation complete! @githubnemo

I’ve added just two attributes to the ```LoraConfig``` class: ```use_arrow``` and ```arrow_config```.
I also refined the Arrow forward pass to reduce VRAM usage.

Currently working on the test suite—will push the final version soon.

githubnemo commented on src/peft/tuners/lora/arrow.py:
```diff
@@ -0,0 +1,136 @@
+import torch
```
```suggestion
# Copyright 2025-present the HuggingFace Inc. team.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import torch
```

TheTahaaa replied:
Done!

githubnemo commented on method_comparison/MetaMathQA/run.py:
```diff
@@ -30,6 +30,7 @@
 from typing import Any, Callable, Literal, Optional
 
 import torch
+from data import get_train_valid_test_datasets
```
Revert?

TheTahaaa replied:
I’m not entirely sure where this came from 🤨
I suspect it’s a result of the previous merge I did with the main branch.

githubnemo reviewed (commented):
Nice! This is already looking good, I think that this design works quite well.

Lots of smaller nit picks which need to be resolved first before doing a more thorough review. One bigger point is the precomputation of prototypes / gks values which disallows adding / removing adapters. It also prevents us from switching to an expert adapter, do fine-tuning and switching back to the router (the prototypes would be outdated). While I think that supporting adding / removing adapters is a smaller change, allowing for fine-tuning while arrow layers are present is a bit more complicated. I expect GKS to be OK with this since we only subtract the weights once at the beginning and could do fine-tuning afterwards. I'm not sure if it is possible to solve this without removing prototype computation. Maybe we should make prototype computation optional. WDYT?

TheTahaaa commented:
> Nice! This is already looking good, I think that this design works quite well.
> 
> Lots of smaller nit picks which need to be resolved first before doing a more thorough review. One bigger point is the precomputation of prototypes / gks values which disallows adding / removing adapters. It also prevents us from switching to an expert adapter, do fine-tuning and switching back to the router (the prototypes would be outdated). While I think that supporting adding / removing adapters is a smaller change, allowing for fine-tuning while arrow layers are present is a bit more complicated. I expect GKS to be OK with this since we only subtract the weights once at the beginning and could do fine-tuning afterwards. I'm not sure if it is possible to solve this without removing prototype computation. Maybe we should make prototype computation optional. WDYT?


The precomputation of prototypes (and the averaging of general experts for GKS) is lazy—it happens in the forward() of ```ArrowLoraLinearLayer``` only the first time the "arrow_router" adapter is active.

This means fine-tuning with multiple experts is fully supported as long as "arrow_router" is not active during training. You can activate other adapters, add/remove experts, fine-tune as needed, and then switch back to "arrow_router". At that moment, prototypes (and GKS values) will be computed based on the currently loaded experts. 

However, in the case of adding or removing LoRA modules to/from the arrow_model, we need to ensure that the task-specific and general LoRA module lists are kept in sync. This can be handled via a utility function (similar to ```.update_layer()```), which would reorganise and refresh the module references before the forward pass when "arrow_router" is active

TheTahaaa commented on src/peft/tuners/lora/config.py:
```diff
@@ -566,6 +599,12 @@ class LoraConfig(PeftConfig):
             )
         },
     )
+    use_arrow: Optional[bool] = field(
+        default=False, metadata={"help": "Whether to apply Arrow routing on the model or not."}
+    )
```
You're right! 👍🏻

TheTahaaa commented on src/peft/tuners/lora/bnb.py:
```diff
@@ -313,9 +322,16 @@ def __init__(
-        def resolve_lora_variant(self, *, use_dora: bool, **kwargs) -> Optional[LoraVariant]:
+        def resolve_lora_variant(self, *, use_dora: bool, use_arrow: bool, **kwargs) -> Optional[LoraVariant]:
+            if use_arrow:
+                # arrow variant must be imported so it registers itself
```
I agree!

TheTahaaa commented on tests/test_arrow.py:
```diff
@@ -0,0 +1,128 @@
+# tests/test_arrow.py
```
Done.

TheTahaaa commented on src/peft/tuners/lora/arrow.py:
```diff
@@ -0,0 +1,136 @@
+                del lora_A[n]
+                del lora_B[n]
+
+            # # 4) register the avg as a new adapter at position num_experts
+            # avg_name = "gks_avg"
+            # device = self.prototypes.device
+            # A_mod = nn.Linear(avg_A.size(1), avg_A.size(0), bias=False).to(device)
+            # B_mod = nn.Linear(avg_B.size(1), avg_B.size(0), bias=False).to(device)
+
+            # A_mod.weight.data.copy_(avg_A.to(device))
+            # B_mod.weight.data.copy_(avg_B.to(device))
+
+            # lora_A[avg_name] = A_mod
+            # lora_B[avg_name] = B_mod
+            # self.expert_names.append(avg_name)
+
+            # # 5) reset caches
+            # E = len(self.expert_names)
+            # new_protos = torch.empty((E, self.in_features), device=self.prototypes.device, dtype=self.prototypes.dtype)
+            # # this replaces the old buffer in-place with newer size
+            # self.prototypes = new_protos
+            # self._protos_ready = False
+            # self.precomputed_W = None
+
+            # print(lora_A)
+            # print(lora_B)
```
Yep, it was for debugging.

TheTahaaa commented on src/peft/tuners/lora/config.py:
```diff
@@ -69,6 +69,39 @@ class LoftQConfig:
+
+    def __post_init__(self):
+        if self.arrow_top_k <= 0:
+            raise ValueError("`arrow_top_k` could not be negative.")
```
Accurate!

TheTahaaa commented on src/peft/tuners/lora/arrow.py:
```diff
@@ -0,0 +1,136 @@
+        self.cast_input_dtype_enabled = True
+
+    @torch.no_grad()
+    def build_prototypes(self, lora_A, lora_B=None):
```
I added a thorough explanation.

TheTahaaa commented:
> > > > However, in the case of adding or removing LoRA modules to/from the arrow_model, we need to ensure that the task-specific and general LoRA module lists are kept in sync. This can be handled via a utility function (similar to .update_layer()), which would reorganise and refresh the module references before the forward pass when "arrow_router" is active
> > > 
> > > 
> > > Yes, something like `update_layer` would work. We don't support this in variants right now but I think adding support shouldn't be doable. For now, let's just add an arrow-specific exception to `update_layers`, something like
> > > ```python
> > > def update_layer(self, adapter_name, ...):
> > >     [...]
> > > 
> > >     if hasattr(self, "lora_arrow"):
> > >         for adapter in self.lora_variant:
> > >             if adapter in self.lora_arrow:
> > >                 self.lora_arrow[adapter].on_adapter_change()
> > > ```
> > 
> > 
> > Should I go ahead and implement `on_adapter_change()` now, or just raise a ValueError for the time being?
> > The point is that if an adapter is loaded or removed from the model, the user should explicitly indicate which adapters are task-specific and which are general adapters used in GKS.
> > One possible approach is to add an attribute in `ArrowLoraLinearLayer`, e.g., `last_adapter_nums`, initialised with the number of task-specific and general LoRAs. Then, when `on_adapter_change()` is called, we can easily check whether the loaded LoRAs have changed.
> > That said, I think that whenever a user adds or removes a LoRA from the _arrow_model_, they should also indicate the final set of task-specific and general LoRAs so we can properly organise things.
> > What do you think?
> 
> Yeah, let's go ahead and implement `on_adapter_change`. I think we can assume the general knowledge adapters to be fixed which is a reasonable assumption since they are by definition task-unspecific and therefore unlikely to be trained or swapped in such a setup. Therefore I don't think we'll need to worry about the user discerning between task-specific/general and assume that every adapter added to the model is task-specific. So `.on_adapter_change` would only need to do GKS and computing prototypes.

✅ Implemented!

```on_adapter_change()``` now checks all adapters currently loaded in the model. If any adapter has been added or removed, it sets the gks and prototype flags to False.
This ensures the model recomputes them once in the forward pass.

TheTahaaa commented on src/peft/tuners/lora/config.py:
```diff
@@ -69,6 +69,39 @@ class LoftQConfig:
+
+    gen_names: list[str] = field(default=None, metadata={"help": "list of general LoRA names."})
+
+    def __post_init__(self):
```
Yes, agree.

TheTahaaa commented on src/peft/tuners/lora/config.py:
```diff
@@ -69,6 +69,39 @@ class LoftQConfig:
+        metadata={"help": "Softmax temperature for computing Arrow expert coefficients."},
+    )
+
+    use_gks: bool = field(
+        default=False,
+        metadata={"help": "Enable GenKnowSub."},
+    )
+
+    ts_names: list[str] = field(default=None, metadata={"help": "list of task-specific LoRA names."})
+
+    gen_names: list[str] = field(default=None, metadata={"help": "list of general LoRA names."})
```
Good catch on the side note!

And regarding the names, I agree.
I changed the ```ts_names``` and ```gen_names``` to ```task_adapter_names``` and ```gks_adapter_names``` respectively.

TheTahaaa commented on src/peft/tuners/lora/arrow.py:
```diff
@@ -0,0 +1,136 @@
+from torch import nn
+
+
+class ArrowLoraLinearLayer(nn.Module):
```
Done.

TheTahaaa commented on src/peft/tuners/lora/arrow.py:
```diff
@@ -0,0 +1,136 @@
+
+    @torch.no_grad()
+    def build_prototypes(self, lora_A, lora_B=None):
+        if self._protos_ready:
+            return
```
It is done by implementing ```on_adapter_change``` inside the ```ArrowLoraLinearLayer``` class.

TheTahaaa commented on src/peft/tuners/lora/arrow.py:
```diff
@@ -0,0 +1,136 @@
+
+    @torch.no_grad()
+    def gen_know_sub(self, lora_A, lora_B):
+        if self.gks_done or not self.use_gks:
+            return
```
It is handled by setting ```self.gks_done = False``` in the ```on_adapter_change()``` .

TheTahaaa commented on src/peft/tuners/lora/arrow.py:
```diff
@@ -0,0 +1,136 @@
+        x : (B, *, in_features) prototypes : (E, in_features) ← built from top-SV of W = B @ A returns : LoRA delta
+        with Arrow routing
+        """
+        # self.prototypes = self.prototypes.to(x.dtype)
```
Yep!

TheTahaaa commented on src/peft/tuners/lora/variants.py:
```diff
@@ -18,13 +18,101 @@
+        if not hasattr(module, "lora_arrow"):
+            module.lora_arrow = nn.ModuleDict()
+            # ensure PEFT knows to save / load these params
+            module.other_param_names = module.other_param_names + ("lora_arrow",)
```
You're right. I removed the saving parameters part, and only kept ```module.lora_arrow = nn.ModuleDict()```

TheTahaaa commented on src/peft/tuners/lora/variants.py:
```diff
@@ -18,13 +18,101 @@
+        )
+        return result + delta
+
+    # ------------------------------------------------------------------
+    # 3) MERGE / UNMERGE  (optional)
+    # ------------------------------------------------------------------
+    # If you plan to support `model.merge_and_unload()`, add:
+    #
+    # @staticmethod
+    # def merge_weights(module: Linear, adapter_name: str) -> None: ...
+    #
+    # For most research prototypes you can skip this block; PEFT will simply
+    # fall back to keeping Arrow in “un-merged” mode.
+
```
Good point, Done!

TheTahaaa commented on src/peft/tuners/lora/variants.py:
```diff
@@ -18,13 +18,101 @@
+
+        delta = arrow(
+            x,
+            lora_A=module.lora_A,  # dict of nn.Linear
```
It’s essentially a ModuleDict() of linear layers, where each layer represents either the A or B matrix in LoRA, correct?

TheTahaaa commented on src/peft/tuners/lora/arrow.py:
```diff
@@ -0,0 +1,136 @@
+        return x.to(dtype=dtype)
+
+    def forward(self, x, lora_A, lora_B, dropout, scaling):
+        """
+        x : (B, *, in_features) prototypes : (E, in_features) ← built from top-SV of W = B @ A returns : LoRA delta
+        with Arrow routing
```
It has been changed to a clean and well-structured description!

TheTahaaa commented on tests/test_arrow.py:
```diff
@@ -0,0 +1,128 @@
+        ).eval()
+
+        x = torch.ones(1, 4, dtype=torch.long)
+        assert not torch.allclose(m_plain(x).logits, m_gks(x).logits)
```
- In ```create_arrow_model()```, I’ve enforced that all task and GKS adapters must target either Linear or Linear4bit layers. If they don’t, a ```TypeError``` is raised. A dedicated test case has been added to verify this behaviour.

- I’ve added safeguards for merging/unmerging when an Arrow variant is active. The test currently covers only the merge scenario, as there’s no existing unmerge test to extend from.

- Added a test to ensure that loading all experts at once via ```create_arrow_model()``` produces the same results as loading all except one in ```create_arrow_model()``` and then adding the remaining one later via ```model.load_adapter()```.

- Added a test for inference with ```torch.float16``` by setting ```autocast_adapter_dtype=False``` and ```torch_dtype=torch.float16``` in create_arrow_model().

TheTahaaa commented:
Hey @githubnemo ,

I’ve updated the prototype computation logic and merged the latest changes from main.
It’s ready for your review! 😁

TheTahaaa commented:
A minor update ensures that all loaded adapters are first checked for consistency in their ```r``` and ```alpha``` values. Once validated, these shared values are then applied to the ```arrow_router```.

githubnemo commented:
I think there was an issue with rebasing and formatting. `src/peft/tuners/__init__.py` is missing the miss and shira changes (there are other places as well where changes from main are reverted, this was just an example). There seems to be a rogue auto-formatter going around as well (there are lots of changes in `method_comparison/` for example).

> There are still some style/formatting issues coming from the merged main branch (see screenshot below), so I haven’t run `make quality` on the merged branch yet:

We've recently updated the ruff version, maybe yours is now outdated. `pip install -e '.[quality]` should update the quality checkers and hopefully resolve the issue.

Could you go over the PR once and try to make sure that the PR aligns with main and doesn't contain irrelevant changes (e.g., auto-formatting of unrelated things). That would help quite a lot, thanks!

TheTahaaa commented:
Hi @githubnemo,

I’ve rebased this branch on the latest main, cleaned up all unrelated changes, and force-pushed a single tidy commit (Arrow tuner + minimal wiring + tests). The branch arrow-genknowsub is now ahead of main by 1 commit.

Could you please reopen this PR so we can continue the review here? 🙏

Thanks a lot!

githubnemo commented on src/peft/tuners/lora/arrow.py:
```diff
@@ -0,0 +1,136 @@
+                lora_A[name].weight.data.sub_(avg_A)
+                lora_B[name].weight.data.sub_(avg_B)
+
+            # 3) now delete the original LE adapters
```
Perfect :)

githubnemo reviewed (commented):
Thanks for dealing with the rebase, everything looks fine now!

There are still some aspects that deserve testing like the hyper-param compatibility check and possibly GPU/bnb related code but in general it looks quite good already.

Since the implementation seems reaching maturity I'd suggest that we focus on adding documentation and examples. You can document the new method in a new section in `docs/source/developer_guides/lora.md` explaining what the method does, that it can be enhanced with GKS and provide brief code examples, similar to the rest of the document. Providing a bit of insight how to manually instantiate the arrow layer for advanced users is a plus. This should include that the naming scheme for loaded adapters is documented (in the code as well) since that is now a major part of how arrow ingests adapters.

In your original post of this PR you mentioned that you already have some examples for common benchmark datasets. Do you think you would be able to contribute an example from one of them? Normally I'd suggest adding the method to the MetaMathQA benchmark suite in PEFT but Arrow (+GKS) make a lot of assumptions that are not true for other methods so I'd rather skip that at this time. Nevertheless, I think having an example that showcases the benefit of Arrow and Arrow + GKS over baseline would be great.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/peft/pr_2644). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

TheTahaaa commented on src/peft/tuners/lora/variants.py:
```diff
@@ -18,13 +18,94 @@
+        """
+        Initialise the ArrowLoraLinearLayer() inside lora_arrow. lora_arrow is nn.ModuleDict(), serving as a container
+        for ArrowLoraLinearLayer(). A layer of the base model with LoRA adapter loaded on it will be like:
+        ----------------------------------------------------
+            (qkv_proj): lora.Linear4bit or lora.Linear(
+                (base_layer): Linear4bit or Linear (lora_dropout): ModuleDict( ... ) (lora_A): ModuleDict( ... )
+                (lora_B): ModuleDict( ... ) (lora_embedding_A): ParameterDict( ... ) (lora_embedding_B): ParameterDict(
+                ... ) (lora_magnitude_vector): ModuleDict( ... ) (lora_arrow): ModuleDict(
+                    (arrow_router): ArrowLoraLinearLayer() )
+            )
+        ----------------------------------------------------
```
I’ve defined the fields in ArrowConfig as follows:
```
task_adapter_names: Optional[list[str]] = field(
        default=None, init=False, metadata={"help": "list of task-specific LoRA adapter names. It will be set in create_arrow_model()."}
    )

    gks_adapter_names: Optional[list[str]] = field(
        default=None, init=False, metadata={"help": "list of general LoRA adapter names for GenKnowSub. It will be set in create_arrow_model()."}
    )
```
This way, users can’t pass values for them directly; they are always set inside create_arrow_model().

On the naming: I initially used expert and adapter interchangeably because of the MoE context (each adapter effectively acts as an expert). I agree, though, that using adapter consistently is clearer. I also like the idea of defining global constants for the prefixes. Right now I’ve placed them inside ```create_arrow_model()```, but I’m open to suggestions on the best location (e.g., ```config.py``` or ```utils/constants.py```).

TheTahaaa commented on src/peft/tuners/lora/arrow.py:
```diff
@@ -0,0 +1,265 @@
+        B32 = B.to(torch.float32)
+        C = B32.T @ B32  # (r, r)
+
+        cpu_state = torch.random.get_rng_state()
+        if torch.cuda.is_available():
+            gpu_state = torch.cuda.get_rng_state()
```
I’ve added an ```rng_seed``` attribute to ```ArrowConfig``` for reproducibility. I also switched the implementation to use a private ```torch.Generator```, which I agree is a cleaner solution. 😁

TheTahaaa commented on src/peft/tuners/lora/arrow.py:
```diff
@@ -0,0 +1,265 @@
+
+class ArrowLoraLinearLayer(nn.Module):
+    """
+    This class represent the main logic of the arrow routing algorithm for linear layers.
```
I simply added comments:
```
self.task_adapter_names = arrow_config.task_adapter_names.copy()   # Set in create_arrow_model() with this format: task_0, task_1, ...
self.gks_adapter_names = arrow_config.gks_adapter_names    # Set in create_arrow_model() with this format: gks_0, gks_1, ...
```

TheTahaaa commented on src/peft/tuners/lora/arrow.py:
```diff
@@ -0,0 +1,265 @@
+        The resulting similarity scores serve as coefficients to compute a weighted average of the corresponding LoRA
+        modules, effectively routing each token through its most relevant experts.
+
+        ** This prototype computation is done only once for all the experts.**
```
Good catch, done!

TheTahaaa commented on tests/test_arrow.py:
```diff
@@ -0,0 +1,318 @@
+# ─── Fixtures ──────────────────────────────────────────────────────────
+
+
+@pytest.fixture(scope="module")
+def base_model():
+    """Tiny GPT-2 variant that ships with HF for unit tests."""
+    return AutoModelForCausalLM.from_pretrained("hf-internal-testing/tiny-random-gpt2")
```
Yep I just removed it.

TheTahaaa commented on tests/test_arrow.py:
```diff
@@ -0,0 +1,318 @@
+
+        # Build a LoRA adapter targeting a Conv2d
+        cfg = LoraConfig(r=4, target_modules=[target_name], init_lora_weights=False)
+        base = AutoModelForImageClassification.from_pretrained(model_id)
```
I did use ```testing_utils.hub_online_once``` to further optimise it.

TheTahaaa commented on tests/test_arrow.py:
```diff
@@ -0,0 +1,318 @@
+        if platform.system() == "Darwin":
+            pytest.skip(reason="MacOS does not support multiple ops in float16")
+
+        device = "cuda" if torch.cuda.is_available() else "cpu"
```
Sorted out!

TheTahaaa commented on src/peft/tuners/lora/variants.py:
```diff
@@ -18,13 +18,94 @@
+        Return value:
+            output of the base model + delta weight computed by arrow layer.
+        """
+        arrow = module.lora_arrow[active_adapter]  # ArrowLoraLinearLayer
+        # Apply GenKnowSub the 1st time if applcable.
+        arrow.gen_know_sub(module.lora_A, module.lora_B)
+        # lazily build prototypes the 1st time after GenKnowSub
+        arrow.build_prototypes(module.lora_A, module.lora_B)
```
Updated.

TheTahaaa commented:
> There are still some aspects that deserve testing like the hyper-param compatibility check and possibly GPU/bnb related code but in general it looks quite good already.
> 
Regarding the bnb concern, I did test both Phi2 and Phi3-mini with 4-bit quantisation on a few datasets like Arc-easy, Arc-challenge, HellaSwag; they work absolutely fine!

> Since the implementation seems reaching maturity I'd suggest that we focus on adding documentation and examples. You can document the new method in a new section in `docs/source/developer_guides/lora.md` explaining what the method does, that it can be enhanced with GKS and provide brief code examples, similar to the rest of the document. Providing a bit of insight how to manually instantiate the arrow layer for advanced users is a plus. This should include that the naming scheme for loaded adapters is documented (in the code as well) since that is now a major part of how arrow ingests adapters.
> 
Sure, but unfortunately seems like there's a problem and I can't login to my HF account:

<img width="467" height="442" alt="Screenshot 2025-08-22 at 5 51 59 PM" src="https://github.com/user-attachments/assets/6f6e7644-571a-466f-b5cd-32e5a2d28397" />


> In your original post of this PR you mentioned that you already have some examples for common benchmark datasets. Do you think you would be able to contribute an example from one of them? Normally I'd suggest adding the method to the MetaMathQA benchmark suite in PEFT but Arrow (+GKS) make a lot of assumptions that are not true for other methods so I'd rather skip that at this time. Nevertheless, I think having an example that showcases the benefit of Arrow and Arrow + GKS over baseline would be great.
Yes, I can. I’ll run some experiments on PIQA, BoolQ, ARC-Easy, etc. to show how GKS and Arrow improve over the base model, using the task adapters trained on the Flan dataset.

TheTahaaa commented:
I’ve added another test to cover the scenario of adding a new adapter, running a forward pass (to build prototype and apply gks), then activating the new adapter and calling forward again (to simulate training/inference with a newly loaded expert), and finally re-activating the arrow_router. The output was identical to the case where all experts were loaded from the start ✅.

In addition, I updated the documentation in ```docs/source/developer_guides/lora.md```. I also prepared a complete code snippet to evaluate the performance of Arrow and GKS against the base model on PIQA, BoolQ, ARC-Easy, etc. I’m just waiting for your confirmation on where to include it — I think the ```examples``` directory would be the right place.

githubnemo commented on src/peft/tuners/lora/variants.py:
```diff
@@ -411,3 +494,258 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+    return remaining if remaining else None
+
+
+def create_arrow_model(
+    base_model: PreTrainedModel,
+    task_specific_adapter_paths: list[str],  # path of task-specific LoRAs
+    ts_repo_id: str,
+    arrow_config: ArrowConfig,
+    general_adapter_paths: list[str] | None = None,  # path to the trained general-knowledge LoRAs
+    gen_repo_id: str | None = None,
+    **adapter_kwargs: Any,
+):
```
After reading the well-written examples, I wonder: Wouldn't it be more user-friendly to just specify `task_specific_adapter_paths` fully qualified and remove the `_repo_id` parameters? I.e.:

```python
create_arrow_model(
    base_model=base_model,
    task_specific_adapter_paths=[
        "TahaBa/phi3-mini-clustered-flan/ts_expert_0",
        "TahaBa/phi3-mini-clustered-flan/ts_expert_1",
        ...
    ],
    [...]
)
```

We could change `get_subfolder_from_path` to something like
this:

```python
def split_model_path(path):
    if os.path.exists(path):
        return path, None
    else:
        split_path = path.split("/")
        model_id = "/".join(split_path[0:2])
        subfolder = "/".join(split_path[2:])
        return model_id, subfolder

# usage
model_id, subfolder = split_model_path(path)

PeftModel.from_pretrained(
    model=base_model,
    model_id=model_id,
    subfolder=subfolder,
)
```

WDYT?

githubnemo commented on docs/source/developer_guides/lora.md:
```diff
@@ -582,3 +582,149 @@ Using this feature has some drawbacks, namely:
+    from transformers import AutoModelForCausalLM, BitsAndBytesConfig
+    import torch
+
+    # Quantisation config
+    bnb_config = BitsAndBytesConfig(
+            load_in_4bit=True,
+            bnb_4bit_quant_type="nf4",
+            bnb_4bit_compute_dtype=torch.bfloat16,
+            bnb_4bit_use_double_quant=False,
+        )
+
+    # Loading the model
+    base_model = AutoModelForCausalLM.from_pretrained(
+        "microsoft/Phi-3-mini-4k-instruct",
+        torch_dtype=torch.bfloat16,
+        device_map="auto",
+        quantization_config=bnb_config,
+    )
+
+    # Now call create_arrow_model() as we explained before.
```
Let's add this as a test case in `test_gpu_examples.py` with a small model like `facebook/opt-125m` and randomly initialized, newly created adapters. Just so that we can see that the model loads and generate works.

githubnemo commented on src/peft/tuners/lora/layer.py:
```diff
@@ -1825,10 +1863,7 @@ def _register_parameter_or_buffer(module, name, X):
 
 class ParamWrapper(nn.Module, LoraLayer):
     """A LoRA wrapper for `nn.Parameter`. This layer is dispatched if users target a parameter directly with
-    `lora_config.target_parameters`
-
-    Note:
-
+    `lora_config.target_parameters` Note:
```
Undo formatting

TheTahaaa replied:
Done.

githubnemo commented on src/peft/tuners/lora/model.py:
```diff
@@ -139,6 +139,9 @@ class LoraModel(BaseTuner):
 
     prefix: str = "lora_"
 
+    def __init__(self, model, config, adapter_name, low_cpu_mem_usage: bool = False) -> None:
+        super().__init__(model, config, adapter_name, low_cpu_mem_usage=low_cpu_mem_usage)
+
```
Remove `__init__`

TheTahaaa replied:
Yep it is the reason for "LoraModel.__init__() got an unexpected keyword argument 'state_dict'" error in CI.

githubnemo reviewed (commented):
Nicely done! Only some minor comments left.

> I think the examples directory would be the right place.

I agree, something like `examples/arrow_multitask` or something akin to the task you have in mind.

Could you also check the CI? There seem to be a few errors due to some LoRA layers still missing the `arrow_config` parameter in their `__init__` and `update_layer` methods.

TheTahaaa commented on tests/test_arrow.py:
```diff
@@ -0,0 +1,389 @@
+        with hub_online_once(model_id):
+            base = AutoModelForImageClassification.from_pretrained(model_id)
+
+        # Find a Conv2d module name in the ResNet test model
+        probe = AutoModelForImageClassification.from_pretrained(model_id)
+        conv_names = [n.split(".")[-1] for n, m in probe.named_modules() if isinstance(m, torch.nn.Conv2d)]
+        assert len(conv_names) > 0, "No Conv2d modules found in the ResNet test model."
```
Yep!

TheTahaaa commented on docs/source/developer_guides/lora.md:
```diff
@@ -582,3 +582,149 @@ Using this feature has some drawbacks, namely:
+
+## Composing and Reusing LoRA Adapters
+### Arrow
+[Arrow](https://arxiv.org/abs/2405.11157) is a modular routing algorithm designed to combine multiple pre-trained task-specific LoRA adapters to solve a given task. Rather than merging all adapters naively, Arrow introduces a **gradient-free, token-wise mixture-of-experts (MoE) routing mechanism**. At inference time, it first computes a _prototype_ for each LoRA by extracting the top right singul ...
```
Done

TheTahaaa commented on docs/source/developer_guides/lora.md:
```diff
@@ -582,3 +582,149 @@ Using this feature has some drawbacks, namely:
+
+# Creating the Arrow config
+arrow_config = ArrowConfig(
+    top_k = 3,
+    router_temperature = 1.0,
+    rng_seed = 42,
```
Good catch!

TheTahaaa commented on docs/source/developer_guides/lora.md:
````diff
@@ -582,3 +582,149 @@ Using this feature has some drawbacks, namely:
+```
+
+### GenKnowSub
+[GenKnowSub](https://aclanthology.org/2025.acl-short.54/) augments Arrow by purifying task-specific LoRA adapters before routing. The key idea is to subtract general knowledge encoded in LoRA space—based on the [forgetting-via-negation principle](https://arxiv.org/abs/2212.04089)—so that task adapters become more isolated and focused on task-relevant signals. Concretely, GenKnowSub estimates a lo ...
````
Corrected.

githubnemo commented on examples/arrow_multitask/arrow_phi3_mini.py:
```diff
@@ -0,0 +1,362 @@
+
+Implementation details:
+- The base model is quantized to 4-bit using `BitsAndBytesConfig` (nf4, bf16 compute).
+- For Arrow and GKS, task-specific adapters are loaded from the Hugging Face Hub:
+    TahaBa/phi3-mini-clustered-flan/ts_expert_i
+- For GKS, general adapters are also loaded from:
+    TahaBa/phi3-mini-general-adapters/...
```
While I think it is OK to use pre-trained adapters and not show-case how they are trained (this is not what arrow / GKS is about), let's add a little bit of info how these are trained (one adapter per task, general knowledge on $lang-wikipedia, etc.).

TheTahaaa replied:
It is added now.

githubnemo commented on docs/source/developer_guides/lora.md:
```diff
@@ -582,3 +582,144 @@ Using this feature has some drawbacks, namely:
+    rng_seed=42,
+)
+
+# The provided LoRA adapters below are trained on a clustered Flan dataset. 
+# The clustering is done using MBC method explained in the Arrow paper.
```
Let's be more specific about what `MBC` is and why that is helpful in the context of arrow.

For example:

> For Arrow it makes sense to cluster tasks based on their similarity and then train one adapter per task cluster (MBC, model-based clustering) to get a number of well-separated adapters. These can then be loaded into the model:

TheTahaaa replied:
I explained the MBC in the ```examples/arrow_multitask/arrow_phi3_mini.py```.

githubnemo commented on docs/source/developer_guides/lora.md:
````diff
@@ -582,3 +582,144 @@ Using this feature has some drawbacks, namely:
+# Now the forward path could be called on this model, like a normal PeftModel.
+```
+
+Furthermore, you can add or remove adapters after calling ```create_arrow_model```—for example, to fine-tune a new adapter or discard an unnecessary one. Once the adapters are in place, you can activate the ```"arrow_router"``` for inference to use Arrow. Note that if you add a new LoRA adapter after ```create_arrow_model``` and want to fine-tune it, you must explicitly set this adapter as active ...
````
```suggestion
Furthermore, you can add or remove adapters after calling ```create_arrow_model```—for example, to fine-tune a new adapter or discard an unnecessary one. Once the adapters are in place, you can activate the ```"arrow_router"``` for inference to use Arrow. Note that if you add a new LoRA adapter after ```create_arrow_model``` and want to fine-tune it, you must explicitly set the new adapter as active, since ```"arrow_router"``` is activated by default in ```create_arrow_model```.
```

TheTahaaa replied:
Good catch!

githubnemo reviewed (commented):
Alright, with tests passing I was able to have a look at the coverage and while there are few tests missing, coverage looks good.

Some smaller comments / suggestions the biggest being moving `create_arrow_model` to `arrow.py` since it has nothing to do with variants as such but nothing too big.

The example looks very clean, thanks for that!

I think once the tests are implemented and CI passes we can merge this.

TheTahaaa commented on tests/test_gpu_examples.py:
```diff
@@ -4970,3 +4974,98 @@ def test_hotswapping_compiled_diffusers_model_does_not_trigger_recompilation(sel
+        yield Path(wd)
+    finally:
+        os.chdir(old_cwd)
+
+
+def _create_and_save_adapter_opt(out_dir: Path, rank: int = 4):
+    """
+    Build a randomly initialized LoRA adapter for OPT-125M and save into `out_dir`. We construct a model from CONFIG
+    (no pretrained weights) to avoid slow downloads here.
+    """
+    model_id = "facebook/opt-125m"
+    # Target all linear layers so the adapter matches whatever we later quantize/load.
+    lora_cfg = LoraConfig(
+        r=rank,
+        lora_alpha=2 * rank,
+        lora_dropout=0.0,
+        bias="none",
+        target_modules="all-linear",
+        task_type="CAUSAL_LM",
+        init_lora_weights=False,  # keep it purely random, as requested
+    )
+    # Create model from config (fast) just to attach/save adapter
+    base_cfg = AutoModelForCausalLM.from_pretrained(model_id, trust_remote_code=False).config
+    base_cfg.tie_word_embeddings = False  # silence warnings for from_config models
+    model = AutoModelForCausalLM.from_config(base_cfg)
+    peft_model = get_peft_model(model, lora_cfg)
+    peft_model.save_pretrained(out_dir)
+
+
+@pytest.fixture(scope="module")
+def ts_adapters_opt(workdir: Path):
+    """
+    Build 3 locally-saved task-specific adapters for OPT-125M and return their absolute paths.
+    """
+    paths = []
+    for i in range(3):
+        sub = workdir / f"ts_expert_{i}"
+        _create_and_save_adapter_opt(sub)
+        paths.append(str(sub))
+    return paths
```
Yep that would be cleaner, done!

TheTahaaa commented on src/peft/tuners/lora/variants.py:
```diff
@@ -411,3 +496,270 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+        import bitsandbytes as bnb  # type: ignore
+
+        Linear4bit = bnb.nn.Linear4bit
+    except Exception:
```
You're right, changed to ```except ImportError```

TheTahaaa commented on src/peft/tuners/lora/variants.py:
```diff
@@ -411,3 +496,270 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+        return model_id, kw
+
+    # Fallback (unlikely): treat as model_id as-is
+    return path, kw
```
Yep no need to change it, resolved.

TheTahaaa commented on src/peft/tuners/lora/layer.py:
```diff
@@ -659,6 +681,16 @@ def merge(self, safe_merge: bool = False, adapter_names: Optional[list[str]] = N
             # no adapter to merge
             return
 
+        # Check for merging in an Arrow model.
+        # Since Arrow is a Mixture-of-Experts (MoE) approach, merging adapters is not
+        # meaningful or even possible: for each token, the top-k LoRA experts are
+        # dynamically selected and routed. Because of this per-token routing, there is
+        # no single set of weights that can represent a merged adapter.
+        if hasattr(self, "lora_arrow"):  # if model is an arrow_model
+            for active_adapter in adapter_names:
+                if active_adapter in self.lora_arrow:  # if lora router was an active adapter
+                    raise RuntimeError("Cannot merge an active Arrow router adapter. Remove it first.")
+
```
Yep this is way cleaner, did it!

TheTahaaa commented on src/peft/tuners/lora/layer.py:
```diff
@@ -712,6 +744,13 @@ def unmerge(self) -> None:
         """
         This method unmerges all merged adapter layers from the base weights.
         """
+
+        # check for unmerging in arrow model
+        if hasattr(self, "lora_arrow"):
+            for name in list(self.merged_adapters):
+                if name in self.lora_arrow:
+                    raise RuntimeError("Cannot unmerge an active Arrow router adapter. Remove it first.")
+
```
Done

TheTahaaa commented on tests/test_arrow.py:
All tests have been added — thanks for pointing these out! The 4th one was especially important.

TheTahaaa commented:
All review suggestions have been addressed in the code. I also verified formatting, and everything is clean now. Ready for your review!

githubnemo commented on tests/test_gpu_examples.py:
```diff
@@ -4970,3 +4974,86 @@ def test_hotswapping_compiled_diffusers_model_does_not_trigger_recompilation(sel
+                device_map="auto",
+                quantization_config=bnb_config,
+            )
+            tok = AutoTokenizer.from_pretrained(model_id, use_fast=True)
```
```suggestion
        with hub_online_once(model_id + 'tokenizer'):
            tok = AutoTokenizer.from_pretrained(model_id, use_fast=True)
```

the tokenizer wasn't fetched on the first invocation of `hub_online_once` so at this point we're already offline and the tokenizer fetching will fail.

You can verify that it works by running

```
pytest -m "single_gpu_tests and bitsandbytes" -k arrow tests/test_gpu_examples.py -x --pdb
```

githubnemo commented on examples/arrow_multitask/arrow_phi3_mini.py:
```diff
@@ -0,0 +1,375 @@
+"""
+This script provides a simple evaluation pipeline for multiple-choice reasoning datasets
+(e.g., BoolQ, HellaSwag, ARC, OpenBookQA, Winogrande) with different composition strategies.
+
+Usage examples:
+    python evaluate.py --strategy base --ds_name arc-challenge
+    python evaluate.py --strategy arrow --ds_name boolq
+    python evaluate.py --strategy gks --ds_name hswag
```
```suggestion

Usage examples:
    python arrow_phi3_mini.py --strategy base --ds_name arc-challenge
    python arrow_phi3_mini.py --strategy arrow --ds_name boolq
    python arrow_phi3_mini.py --strategy gks --ds_name hswag
```

githubnemo commented on examples/arrow_multitask/arrow_phi3_mini.py:
Let's add an `requirements.txt` file with the non-PEFT dependencies (like sklearn).

githubnemo reviewed (commented):
Nice, I'm letting the CI run to see if that passes but except for some nitpicks and a small bug in the GPU test it looks fine.

Regarding docs, let's add `ArrowConfig` to `docs/source/package_reference/lora.md` so that it is discoverable there as well.

Examples seem to run fine as well.

Could you merge with main and resolve the conflicts? If that diff looks good we can merge this PR :)

TheTahaaa commented:
All review suggestions have been applied. I also merged the branch with main and ran the test suite; things look good overall.

While running ```tests/test_lora_variants.py```, I’m seeing warnings that appear to be specific to aLoRA (not Arrow-related):
```
====================================================================== warnings summary ======================================================================
tests/test_lora_variants.py::TestLoraVariants::test_variant_is_applied_to_layers[alora-LoraConfig-config_kwargs1]
tests/test_lora_variants.py::TestActivatedLora::test_calculate_alora_offsets[input_ids0-alora_invocation_tokens0-expected_offsets0]
tests/test_lora_variants.py::TestActivatedLora::test_calculate_alora_offsets[input_ids1-alora_invocation_tokens1-expected_offsets1]
tests/test_lora_variants.py::TestActivatedLora::test_calculate_alora_offsets[input_ids2-alora_invocation_tokens2-expected_offsets2]
tests/test_lora_variants.py::TestActivatedLora::test_calculate_alora_offsets_with_adapter_names[input_ids0-alora_invocations0-expected_offsets0]
tests/test_lora_variants.py::TestActivatedLora::test_calculate_alora_offsets_with_adapter_names[input_ids1-alora_invocations1-expected_offsets1]
tests/test_lora_variants.py::TestActivatedLora::test_alora_activation_matches_base_until_invocation
tests/test_lora_variants.py::TestActivatedLora::test_input_embeds_warning
tests/test_lora_variants.py::TestActivatedLora::test_num_beams_error
  peft/src/peft/tuners/lora/config.py:741: UserWarning: aLoRA is currently only supported for CAUSAL_LM task.
    warnings.warn("aLoRA is currently only supported for CAUSAL_LM task.")
```
These warnings originate from the aLoRA config guard and are expected for non-CAUSAL_LM task types; they are not caused by the Arrow changes. 

The final version with all necessary changes is pushed and ready for review.
